bmoauth.php

```diff
@@ -0,0 +1,242 @@
+    $csrf = $controller->getExtraURIData();
+
+    $token = id(new PhabricatorAuthTemporaryTokenQuery())
+      ->setViewer(PhabricatorUser::getOmnipotentUser())
+      ->withTokenResources(array($provided_trans_code))
+      ->withTokenCodes(array($csrf))
+      ->withTokenTypes(array(self::TOKEN_TYPE))
+      ->withExpired(false)
+      ->executeOne();
+
+    if(!$token) {
+      throw new Exception(pht('No temporary token found for this transaction code (%s) or CSRF token (%s).', $provided_trans_code, $csrf));
+    }
+
+    // Compare result token and client_api_login to original response
+    $token_props = $token->getProperties();
+    if($token_props['client_login'] != $provided_api_login) {
+      throw new Exception(pht('Token\'s API Login does not match Bugzilla API Login.'));
+    }
+
+    // Call BMO Who Am I REST resource to validate API Key (X-Bugzilla-API-Key)
+    $future = id(
+        new HTTPSFuture('http://'.$adapter->getAdapterDomain().'/rest/whoami')
+      )
+      ->setMethod('GET')
+      ->addHeader('X-Bugzilla-API-Key', $token_props['api_key'])
+      ->addHeader('Accept', 'application/json')
+      ->setTimeout(5);
+
+    $whoami_body = '';
+    try {
+      list($whoami_body) = $future->resolvex();
+    } catch (HTTPFutureResponseStatus $ex) {
+      throw new Exception(pht('Bugzilla WhoAmI request failed.'));
+    }
+
+    $user_json = array();
+    try {
+      $user_json = phutil_json_decode($whoami_body);
+    }
+    catch(Exception $e) {
+      throw new Exception(pht('JSON from Bugzilla WhoAmI could not be parsed: '.$whoami_body));
+    }
+
+    // Clean up! Delete temporary token used for this login
+    $unguarded = AphrontWriteGuard::beginScopedUnguardedWrites();
+    $token->delete();
+    unset($unguarded);
+
+    // Save the user information, then creating or loading the account
+    // Example: Array ( [real_name] => Vagrant User [name] => vagrant@bmo-web.vm [id] => 1 )
+    $real_name = $user_json['real_name'];
+    $bugzilla_nick = $adapter->parseBugzillaNick($real_name);
+    if($bugzilla_nick) {
+      $real_name = trim(str_replace(':'.$bugzilla_nick, '', $real_name));
+    }
+
+    $adapter->setAccountName($bugzilla_nick ?: str_replace(' ', '', $real_name));
+    $adapter->setAccountRealName($real_name);
+    $adapter->setAccountID($user_json['id']);
+    $adapter->setAccountEmail($user_json['name']);
+
+    return array($this->loadOrCreateAccount($user_json['id']), $response);
+  }
+}
+
+final class PhabricatorBMOAuthAdapter extends PhutilAuthAdapter {
+
+  private $account_id;
+  private $account_email;
+  private $account_name;
+  private $account_real_name;
+
+  public function getAccountID() {
+    return $this->account_id;
+  }
+
+  public function setAccountId($id) {
+    $this->account_id = $id;
+  }
+
+  public function getAdapterType() {
+    return 'bmo';
+  }
+
+  public function getAdapterDomain() {
+    // return 'bugzilla.mozilla.org';
+    return 'bmo-web.vm';
```

Comment:
We discussed making this a configurable field (via the UI).  Would probably be a good idea.